url8012c8d031_dmbb966_BBMS_tgz-pJ8-testbed_JNEATtestJ8/src/jneat/Genome.java

```diff
@@ -0,0 +1,1528 @@
+		else avgGene.lnk.in_node = geneB.lnk.in_node;
+		
+		if (GlobalFuncs.randFloat() > 0.5) avgGene.lnk.out_node = geneA.lnk.out_node;
+		else avgGene.lnk.out_node = geneB.lnk.out_node;
+		
+		if (GlobalFuncs.randFloat() > 0.5) avgGene.lnk.recurrent = geneA.lnk.recurrent;
+		else avgGene.lnk.recurrent = geneB.lnk.recurrent;
+		
+		avgGene.innovation_num = geneA.innovation_num;
+		avgGene.mutation_num = (geneA.mutation_num + geneB.mutation_num) / 2.0;
+
+		return avgGene;
+	}
+	
+	/** Checks the chosenGene versus the newGene vector and returns if it is duplicate, i.e. should skip
+	 * 
+	 * @param newGenes
+	 * @param chosenGene
+	 * @return
+	 */
+	public boolean CheckGeneConflict(Vector<Gene> newGenes, Gene chosenGene) {
+		boolean skipGene = false;
+		
+		// Check to see if the chosenGene conflicts with one already chosen
+		// i.e. do they represent the same link?
+		
+		Iterator<Gene> itr_newGenes = newGenes.iterator();
+		while (itr_newGenes.hasNext()) {
+			Gene _curGene = itr_newGenes.next();
+			
+			if (_curGene.lnk.in_node.id == chosenGene.lnk.in_node.id &&
+				_curGene.lnk.out_node.id == chosenGene.lnk.out_node.id &&
+				_curGene.lnk.recurrent == chosenGene.lnk.recurrent) {
+				skipGene = true;
+				break;
+			}
+		}		
+		
+		return skipGene;
+	}
+	
+	/** Adds chosenGene and associated nodes and traits 
+	 */
+	public void AddGene(Vector<NNode> newNodes, Vector<Trait> newTraits, Vector<Gene> newGenes, Gene chosenGene, boolean disableGene) {
+		
+		// IF traits are going to be implemented (which they aren't yet) fix this code:
+		// int traitNum = 0;
+		
+		// Add chosenGene to the child
+		// if (chosenGene.lnk.linkTrait == null) traitNum = traits.firstElement().id;
+		// else traitNum = chosenGene.lnk.linkTrait.id - traits.firstElement().id; 
+		
+		NNode inode = chosenGene.lnk.in_node;
+		NNode onode = chosenGene.lnk.out_node;
+		NNode newinode = null;
+		NNode newonode = null;
+		
+		boolean foundiNode = false;
+		boolean foundoNode = false;
+		
+		// For ordering, and stuff.
+		if (inode.id >= onode.id){
+			NNode temp = inode;
+			inode = onode;
+			onode = temp;
+		}
+						
+		// Search the inode and onode
+		for (int i = 0; i < newNodes.size(); i++) {
+			NNode curNode = newNodes.elementAt(i);
+			if (curNode.id == inode.id){
+				foundiNode = true;
+				newinode = curNode;							
+			}
+			if (curNode.id == onode.id) {
+				foundoNode = true;
+				newonode = curNode;
+			}
+		}
+		
+		// Insert inode if needed
+		if (!foundiNode) {
+			System.out.println("DEBUG: Inserting iNode");
+			
+			int nodeTraitNum = 0;
+			
+			if (inode.nodeTrait != null) nodeTraitNum = inode.nodeTrait.id - traits.firstElement().id;						
+			
+			//Trait newTrait = newTraits.elementAt(nodeTraitNum);
+			Trait newTrait = new Trait();
+			newinode = new NNode(inode, newTrait);
+			System.out.println("DEBUG: Inserted node is: " + newinode.PrintNode());
+			node_insert(newNodes, newinode);						
+		}				
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 731:743, 746:758.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.